src/v/cluster/tm_stm.cc

```diff
@@ -680,51 +791,91 @@ ss::future<tm_stm::op_status> tm_stm::add_group(
     co_return tm_stm::op_status::success;
 }
 
+bool tm_stm::check_tx_membership(const kafka::transactional_id& tx_id) {
+    return _hosted_txes.contains(tx_id);
+}
+
 ss::future<>
 tm_stm::apply_snapshot(stm_snapshot_header hdr, iobuf&& tm_ss_buf) {
     vassert(
-      hdr.version == supported_version,
+      hdr.version >= tm_snapshot_v0::version
+        && hdr.version <= tm_snapshot::version,
       "unsupported seq_snapshot_header version {}",
       hdr.version);
     iobuf_parser data_parser(std::move(tm_ss_buf));
-    auto data = reflection::adl<tm_snapshot>{}.from(data_parser);
+    if (hdr.version == tm_snapshot_v0::version) {
+        auto data = reflection::adl<tm_snapshot_v0>{}.from(data_parser);
+
+        _cache->clear_mem();
+        _cache->clear_log();
+        for (auto& entry : data.transactions) {
+            _cache->set_log(entry);
+            _pid_tx_id[entry.pid] = entry.id;
+        }
+        _last_snapshot_offset = data.offset;
+        _insync_offset = data.offset;
+    } else if (hdr.version == tm_snapshot::version) {
+        auto data = reflection::adl<tm_snapshot>{}.from(data_parser);
 
-    _cache->clear_mem();
-    _cache->clear_log();
-    for (auto& entry : data.transactions) {
-        _cache->set_log(entry);
-        _pid_tx_id[entry.pid] = entry.id;
+        _cache->clear_mem();
+        _cache->clear_log();
+        for (auto& entry : data.transactions) {
+            _cache->set_log(entry);
+            _pid_tx_id[entry.pid] = entry.id;
+        }
+        _last_snapshot_offset = data.offset;
+        _insync_offset = data.offset;
+        _hosted_txes = std::move(data.hash_ranges);
+        _hosted_txes.inited = true;
     }
-    _last_snapshot_offset = data.offset;
-    _insync_offset = data.offset;
 
     return ss::now();
 }
 
 ss::future<stm_snapshot> tm_stm::take_snapshot() {
-    return ss::with_gate(_gate, [this] { return do_take_snapshot(); });
+    // Update hash ranges to always have batch in log
+    // So it cannot be deleted with cleanup policy
+    auto update_hash_ranges_res = co_await update_hosted_transactions(
+      _insync_term, _hosted_txes);
+    if (update_hash_ranges_res) {
```

Comment:
please do explicit checks otherwise a reviewer needs to jump to definition of `tm_stm::op_status`